1-Draft/RFC00XX-Update-Error-View.md

```diff
@@ -0,0 +1,181 @@
+---
+RFC: Update $ErrorView with Single String Error Messages
+Author: Jason Helmick
+Status: Draft
+Version: 0.0.0
+Area: PowerShell
+Comments Due: 
+---
+
+# Update $ErrorView with Single String Error Messages
+
+When an error occurs in PowerShell, the customers on-screen error message experience currently
+provides a level of detail that obscures the inner exception message from being read by
+new or occasional PowerShell users.
+To improve both comprehension and the troubleshooting experience, a single string of text
+containing the error will be displayed.  A cmdlet will provide the full error details when desired.
+
+## Motivation
+
+The on-screen experience, when receiving an error message,
+is controlled through the views NormalView (the default) and CategoryView. These are user selectable though the preference variable $ErrorView.
+The addition of a “ConciseView” as default that contains only specific error
+information producing a single line on screen and in logs will improve customer success.
+
+Comprehensive error information is available to the customer through
+accessing the error object. For new and occasional customers,
+a cmdlet “Resolve-ErrorRecord” will simplify the process of getting detailed error information, improving customers success.
+
+## Specification
+
+The proposal is to add a new single-line default error message view called ConciseView
+and add a new cmdlet (Resolve-ErrorRecord) to provide detailed error information.
+
+__Key Design Considerations__
+
+1. To reduce confusion and improve debugging success,
+error messages by default should produce a single line of text, including the word “ERROR:” to make consistent with Verbose and Warning messages. This is added to $errorView as view named "ConciseView".
+
+- $ErrorView should contain these views
+
+    + NormalView
+    + CategoryView
+    + ConciseView
+
+- Error Message syntax for ConciseView
+
+```powershell
+ERROR: <Exception Message>: <Positional Info>
+ERROR: Cannot find path ‘C:\blah’ because it does not exist: At line:1 char:1
+```
+
+2. A new cmdlet “Resolve-ErrorRecord” will produce detailed error information similar to Format-List * -Force.
```

Comment:
This command should be called `Format-Error` ... since that's what it does.

Reply:
Just as a note of reference - the AZ commands have a Resolve-Error cmdlet.  I agree with Jaykul that Format-Error is probably the better verb.

Reply:
If you're to have a new cmdlet, `Format` wouldn't be the better verb. `Format` cmdlets are to tell PowerShell how to format data that is passed into it.

That said, _you really shouldn't add a new cmdlet here, unless you're hearing that `$error` is too complicated_. If that's really the case, I recommend `Get-Error` over any of the other alternatives.

The points made in [these comments](https://github.com/PowerShell/PowerShell-RFC/pull/215#discussion_r309509455) explain how I think it make more sense to approach this.

Reply:
I think `$error` has discovery issues and finding your error in the `$error` collection is problematic.  As soon you tell folks, oh, just do `$error[0] | Format-Error -Detailed`, they' ll type it in incorrectly: `$errror[0] | Format-Error -Detailed`.  That'll generate a new error so by the time they correct their syntax, that command is no longer returning details on the error the user is interested in.

I think the output of this new concise view should include a small snippet of text that instructs users on how to get more detailed error info on **this specific error**.  

Given that you can't really rely on an index in `$error`, I've suggested before that maybe the Concise error displays a 7/8 digit SHA1 hash that can be used to access this error irrespective of its current index in `$error`.  So maybe the extra error text looks like this:
```
08-20 22:45:28 15ms 10> gci xyzzy
gci : Cannot find path 'C:\Users\Keith\xyzzy' because it does not exist.
For more details, run Resolve-Record a64e5e07
```

Reply:
Ok, that's a great point, and after working in PS for many years, it's easy to forget that pain point (trying to find the right error index when you introduce additional errors) when you instinctively deal with it all the time.

I don't like the idea of a 7-8 digit SHA1 though -- too many characters to type, and pushes for copy-paste too much. What if we captured the history id for each error, that is generated/stored in `$error`? I like that better for a bunch of reasons:

1. I'm looking at HistoryPx and thinking of how I can bring it forward into PowerShell 7, and capturing the history for each captured error would be useful for that.
1. I could pull up any captured errors that haven't been cleared for a specific command using syntax like: `Get-Error -HistoryId 2`, or `Get-Error -HistoryId 2 -First 3`, where `-First` speaks to the order in which the errors occurred, which is very relevant when debugging/troubleshooting -- much more so than looking at the last error.
1. I could get errors for the last command I ran at any time using `Get-Error -HistoryId -1` (or `Get-Error -LastCommand`).
1. If `Get-Error` grouped output by HistoryId (maybe showing the command that was invoked even), it would be easier to retrieve the right error in a subsequent invocation because you'd see the Id to work with in the output.
1. `Resolve-Error` could simply accept any `ErrorRecord` as piped input, leaving the selection process to decide which errors you care about up to `Get-Error`, and the heavy lifting to dig deeper into `Exception` and `InnerException`(s) up to `Resolve-Error`.

We could also make it easier to manage errors without throwing them all away by having additional commands such as `Clear-Error -HistoryId 2..5`, or perhaps add a `-Error` switch to `Clear-History` that removes corresponding errors from the `$error` cache as well (an approach I really like, at least at first glance).

Reply:
>   What if we captured the history id for each error, that is generated/stored in `$error`? I like that better ....

>  ..... I could get errors for the last command I ran at any time using `Get-Error -HistoryId -1` (or `Get-Error -LastCommand`).

This should be trivial to do because the Error Record contains Invocation info but there is a **bug** where the history ID is always -1  But $error.where({$_.invocationInfo.HistoryID = 42}) should do it today. 
The putative get-error might well default to "last command". if you have to enter one, simple ID good: GUID (for example) bad 

> I also wonder if the default `prompt` in PowerShell shouldn't show the history id of the command you're about to execute as well...that would make all of this really easy to work with.

**Yes**. Many people reading this will have set this for themselves, and most will have had  someone ask "how do you do that". Making it easier to use #<id>[tab] etc is definitely good.

Reply:
@KirkMunro The SHA1 was just an example of one way to get a semi-permanent id for the error.  Any approach that accomplishes that is good by me.  :-)

Reply:
I think it would be interesting to have the error stream as a PSProvider/PSDrive, so you could use existing cmdlets to navigate the records like a file system:

dir error:\ | select -first 1
dir error:\ | ? id -eq 0 | select *
dir error:\ | ? id -eq 0 | select Exception, FullyQualifiedErrorId

Not sure how difficult that would be to implement, but just a thought.

Reply:
> This should be trivial to do because the Error Record contains Invocation info but there is a **bug** where the history ID is always -1

@jhoneill When you have time, can you share the scenarios where you run into that as a bug?

There are only two scenarios I can recall where `HistoryId` is set to -1 in the `InvocationInfo` in an `ErrorRecord`:

1. When you invoke commands inside of a debugger that generate errors. This is expected, because commands invoked within a debugger are not added to history. Programmatically there are other ways you can invoke commands that are not added to history, and any of those would result in a `HistoryId` of -1.

1. Certain processing errors. For example: `& {[CmdletBinding(InvalidArg=$true)]param(); 'What?'}`. That command will show up in history, but the error that it produces will have a history ID of -1. That's a bug IMHO.

I know there are more, and it would be good to identify them so that they can be fixed.

Reply:
@KirkMunro  
It appears the the invocation info always gives the current ID

I wrote these
```
function invoke-error {
    $MyInvocation.HistoryId
    1/0
}

function Get-error {
    $MyInvocation.HistoryId
    $Error[0].invocation.historyid
}


```

And ran them 
```
#505 PS5 C:\...Powershell>invoke-error
505
Attempted to divide by zero.
At line:3 char:5

#506 PS5 C:\...Powershell>$Error[0].InvocationInfo.HistoryId
-1

#507 PS5 C:\....Powershell>Get-error
507
```
So it looks like it always returns the **current** history id in a function and -1 at the prompt. 

I think you've added two more cases. :-)

Reply:
Thanks, I've moved this information into [PowerShell/PowerShell #10467](https://github.com/PowerShell/PowerShell/issues/10467) so that we can track and fix it.

Reply:
This thread is marked as resolved, but it doesn't feel like some of the thoughts shared were responded to, nor were there corresponding changes in the RFC.